### ISSUE TITLE
dr-loss reimplements following guides

### ISSUE THREAD

**Issue by mytk2012** [open-mmlab/mmdetection#3903]
**error:**
![image](https://user-images.githubusercontent.com/57787557/95650136-2270c780-0b14-11eb-8b2c-765340cfe6a1.png)
**code**
import torch
from torch import nn
import torch.nn.functional as F
import math
from ..builder import LOSSES
from .utils import weight_reduce_loss
"""
    PyTorch Implementation for DR Loss
    Reference
    CVPR'20: "DR Loss: Improving Object Detection by Distributional Ranking"
    Copyright@Alibaba Group Holding Limited
"""
@weight_reduce_loss
def sigmoddrloss(pred, target,weight,neg_lambda,pos_lambda,loss_weight,L,margin,reduction='mean',avg_factor=None):
    logits = torch.abs(pred - target)
    num_classes = logits.shape[1]
    dtype = target.dtype
    device = target.device
    class_range = torch.arange(1, num_classes + 1, dtype=dtype, device=device).unsqueeze(0)
    t = target.unsqueeze(1)

    pos_ind = (t == class_range)
    neg_ind = (t != class_range) * (t >= 0)

    pos_prob = logits[pos_ind].sigmoid()
    neg_prob = logits[neg_ind].sigmoid()

    neg_q = F.softmax(neg_prob /neg_lambda, dim=0)  # lambda是λ
    neg_dist = torch.sum(neg_q * neg_prob)

    if pos_prob.numel() > 0:
        pos_q = F.softmax(-pos_prob /pos_lambda, dim=0)
        pos_dist = torch.sum(pos_q * pos_prob)

        loss_c = loss_weight * torch.log(1. + torch.exp(L * (neg_dist - pos_dist + margin))) / L
    else:

        loss_c = loss_weight * torch.log(1. + torch.exp(L * (neg_dist - 1. + margin))) / L
    if weight is not None:
        if weight.shape != loss_c.shape:
            if weight.size(0) == loss_c.size(0):
                # For most cases, weight is of shape (num_priors, ),
                #  which means it does not have the second axis num_class
                weight = weight.view(-1, 1)
            else:
                # Sometimes, weight per anchor per class is also needed. e.g.
                #  in FSAF. But it may be flattened of shape
                #  (num_priors x num_class, ), while loss is still of shape
                #  (num_priors, num_class).
                assert weight.numel() == loss_c.numel()
                weight = weight.view(loss_c.size(0), -1)
        assert weight.ndim == loss_c.ndim
    loss_cl= weight_reduce_loss(loss_c, weight, reduction, avg_factor)
    return loss_cl

@LOSSES.register_module()
class SigmoidDRLoss(nn.Module):

    def __init__(self, use_sigmoid=True,pos_lambda=1, neg_lambda=0.1/math.log(3.5), L=6., reduction='mean',loss_weight=4.):
        super(SigmoidDRLoss, self).__init__()
        self.use_sigmoid = use_sigmoid
        self.margin = 0.5
        self.pos_lambda = pos_lambda
        self.neg_lambda = neg_lambda
        self.L = L
        self.loss_weight = loss_weight
        self.reduction=reduction

    def forward(self, pred, target,weight=None,avg_factor=None,reduction_override=None):

        assert reduction_override in (None, 'none', 'mean', 'sum')
        reduction = (
            reduction_override if reduction_override else self.reduction)
        loss_cls = self.loss_weight * sigmoddrloss(
            pred,
            target,
            weight,
            neg_lambda=self.neg_lambda,
            pos_lambda=self.pos_lambda,
            loss_weight=self.loss_weight,
            L=self.L,
            margin=self.margin,
            reduction=reduction,
            avg_factor=avg_factor)
        return loss_cls

**Comment by ZwwWayne** [comment]
It seems that you should use`weighted_loss` rather than `weight_reduce_loss` as a wrapper, see example [here](https://github.com/open-mmlab/mmdetection/tree/master/mmdet/models/losses#L8). Further more, it seems that your loss already uses `weight_reduce_loss` in the implementation, so you may check whether the wrapper is necessary.

**Comment by mytk2012** [comment]
> It seems that you should use`weighted_loss` rather than `weight_reduce_loss` as a wrapper, see example [here](https://github.com/open-mmlab/mmdetection/tree/master/mmdet/models/losses#L8). Further more, it seems that your loss already uses `weight_reduce_loss` in the implementation, so you may check whether the wrapper is necessary.

Can you give me some suggestions about drloss code applied in mmd?

**Comment by ZwwWayne** [comment]
I think I have already given. You may also refer to the [G-Focal Loss implementation](https://github.com/open-mmlab/mmdetection/blob/master/mmdet/models/losses/gaussian_focal_loss.py) as an example.
